dali/base/dasds.cpp

```diff
@@ -5695,49 +5699,113 @@ class CStoreHelper : implements IStoreHelper, public CInterface
+        try
+        {
+            OwnedIFileIOStream fstream = createIOStream(iFileIOTmpStore);
+            Owned<ICrcIOStream> crcPipeStream = createCrcPipeStream(fstream);
+            Owned<IIOStream> ios = createBufferedIOStream(crcPipeStream);
 
-        unsigned edition = storeInfo.edition;
-        unsigned newEdition = nextEditionN(edition);
-        bool done = false;
+#ifdef _DEBUG
+            toXML(root, *ios); // formatted (default)
+#else
+            toXML(root, *ios, 0, 0);
+#endif
+            ios->flush(); // ensure flushed outside of dtor (to ensure any exception thrown)
+            ios.clear();
+            fstream.clear();
+            crcPipeStream->flush(); // ensure flushed outside of dtor. NB: calls wrapped stream flush()
+            crc = crcPipeStream->queryCrc();
+            crcPipeStream.clear();
+            iFileIOTmpStore->close(); // ensure flushed outside of dtor
+            iFileIOTmpStore.clear();
+        }
+        catch (IException *e)
+        {
+            DISLOG(e, "Exception - Error saving store XML file");
+            iFileIOTmpStore.clear();
+            iFileTmpStore->remove();
+            throw;
+        }
+
+        return crc;
+    }
+    unsigned saveStoreToBinaryFile(IPropertyTree *root)
+    {
+        LOG(MCdebugProgress, "Saving store to binary");
+        assertex(root);
+        unsigned crc{0};
+
+        MemoryBuffer memBuf;
+        StringBuffer tmpStoreName;
         try
         {
-            unsigned crc = 0;
-            StringBuffer tmpStoreName;
-            OwnedIFileIO iFileIOTmpStore = createUniqueFile(location, TMPSAVENAME, NULL, tmpStoreName);
+            OwnedIFileIO iFileIOTmpStore = createUniqueFile(location, temporaryBinaryFileSaveName, NULL, tmpStoreName);
             OwnedIFile iFileTmpStore = createIFile(tmpStoreName);
             try
             {
+                // Serialize to memory buffer first to enable CRC calculation
+                Owned<IBufferedSerialOutputStream> out = createBufferedSerialOutputStream(memBuf);
+                root->serializeToStream(*out);
+                out->flush();
+                out.clear();
+
+                // Write to file through CRC stream
                 OwnedIFileIOStream fstream = createIOStream(iFileIOTmpStore);
                 Owned<ICrcIOStream> crcPipeStream = createCrcPipeStream(fstream);
-                Owned<IIOStream> ios = createBufferedIOStream(crcPipeStream);
-
-#ifdef _DEBUG
-                toXML(root, *ios);          // formatted (default)
-#else
-                toXML(root, *ios, 0, 0);
-#endif
-                ios->flush(); // ensure flushed outside of dtor (to ensure any exception thrown)
-                ios.clear();
-                fstream.clear();
-                crcPipeStream->flush(); // ensure flushed outside of dtor. NB: calls wrapped stream flush()
+                crcPipeStream->write(memBuf.length(), memBuf.toByteArray());
+                crcPipeStream->flush();
```

Comment:
eek, writing whole thing to a MB, to read it back again, purely to get a crc..

The way to do this, would be to implement a IBufferedSerialOutputStream wrapper, that passes through, but crc's the byte stream it sees, much in the same way that the XML write does it.

I do wonder whether it's worth CRC's at all,
but the data is critical, so maybe, and impl. should be quite simple. Let's keep it, but make it configurable on an expert option.